File: python/semantic_kernel/core_skills/text_memory_skill.py
# Copyright (c) Microsoft. All rights reserved.
import json
import typing as t

from semantic_kernel.sk_pydantic import PydanticField
from semantic_kernel.skill_definition import sk_function, sk_function_context_parameter

if t.TYPE_CHECKING:
    from semantic_kernel.orchestration.sk_context import SKContext


class TextMemorySkill(PydanticField):
    COLLECTION_PARAM = "collection"
    RELEVANCE_PARAM = "relevance"
    KEY_PARAM = "key"
    LIMIT_PARAM = "limit"
    DEFAULT_COLLECTION = "generic"
    DEFAULT_RELEVANCE = 0.75
    DEFAULT_LIMIT = 1

    # @staticmethod
    @sk_function(
        description="Recall a fact from the long term memory",
        name="recall",
        input_description="The information to retrieve",
    )
    @sk_function_context_parameter(
        name=COLLECTION_PARAM,
        description="The collection to search for information",
        default_value=DEFAULT_COLLECTION,
    )
    @sk_function_context_parameter(
        name=RELEVANCE_PARAM,
        description="The relevance score, from 0.0 to 1.0; 1.0 means perfect match",
        default_value=DEFAULT_RELEVANCE,
    )
    @sk_function_context_parameter(
        name=LIMIT_PARAM,
        description="The maximum number of relevant memories to recall.",
        default_value=DEFAULT_LIMIT,
    )
    async def recall_async(self, ask: str, context: "SKContext") -> str:
        """
        Recall a fact from the long term memory.

        Example:
            sk_context["input"] = "what is the capital of France?"
            {{memory.recall $input}} => "Paris"

        Args:
            ask -- The question to ask the memory
            context -- Contains the 'collection' to search for information
                , the 'relevance' score to use when searching
                and the 'limit' of relevant memories to retrieve.

        Returns:
            The nearest item from the memory store as a string or empty string if not found.
        """
        if context.variables is None:
            raise ValueError("Context has no variables")
        if context.memory is None:
            raise ValueError("Context has no memory")

        collection = (
            context.variables[TextMemorySkill.COLLECTION_PARAM]
            if context.variables.contains_key(TextMemorySkill.COLLECTION_PARAM)
            else TextMemorySkill.DEFAULT_COLLECTION
        )
        if not collection:
            raise ValueError("Memory collection not defined for TextMemorySkill")

        relevance = (
            context.variables[TextMemorySkill.RELEVANCE_PARAM]
            if context.variables.contains_key(TextMemorySkill.RELEVANCE_PARAM)
            else TextMemorySkill.DEFAULT_RELEVANCE
        )
        if relevance is None or str(relevance).strip() == "":
            relevance = TextMemorySkill.DEFAULT_RELEVANCE

        limit = (
            context.variables[TextMemorySkill.LIMIT_PARAM]
            if context.variables.contains_key(TextMemorySkill.LIMIT_PARAM)
            else TextMemorySkill.DEFAULT_LIMIT
        )
        if limit is None or str(limit).strip() == "":
            limit = TextMemorySkill.DEFAULT_LIMIT

        results = await context.memory.search_async(
            collection=collection,
            query=ask,
            limit=int(limit),
            min_relevance_score=float(relevance),
        )
        if results is None or len(results) == 0:
            if context.log is not None:
                context.log.warning(f"Memory not found in collection: {collection}")
            return ""
        return results[0].text if limit == 1 else json.dumps([r.text for r in results])

    @sk_function(
        description="Save information to semantic memory",
        name="save",
        input_description="The information to save",
    )
    @sk_function_context_parameter(
        name=COLLECTION_PARAM,
        description="The collection to save the information",
        default_value=DEFAULT_COLLECTION,
    )
    @sk_function_context_parameter(
        name=KEY_PARAM,
        description="The unique key to associate with the information",
    )
    async def save_async(self, text: str, context: "SKContext"):
        """
        Save a fact to the long term memory.

        Example:
            sk_context["input"] = "the capital of France is Paris"
            sk_context[TextMemorySkill.KEY_PARAM] = "countryInfo1"
            {{memory.save $input}}

        Args:
            text -- The text to save to the memory
            context -- Contains the 'collection' to save the information
                and unique 'key' to associate with the information
        """
        if context.variables is None:
            raise ValueError("Context has no variables")
        if context.memory is None:
            raise ValueError("Context has no memory")

        collection = (
            context.variables[TextMemorySkill.COLLECTION_PARAM]
            if context.variables.contains_key(TextMemorySkill.COLLECTION_PARAM)
            else TextMemorySkill.DEFAULT_COLLECTION
        )
        if not collection:
            raise ValueError("Memory collection not defined for TextMemorySkill")

        key = (
            context.variables[TextMemorySkill.KEY_PARAM]
            if context.variables.contains_key(TextMemorySkill.KEY_PARAM)
            else None
        )
        if not key:
            raise ValueError("Memory key not defined for TextMemorySkill")

        await context.memory.save_information_async(collection, text=text, id=key)

File: java/plugins/semantickernel-plugin-core/src/main/java/com/microsoft/semantickernel/coreskills/TextMemorySkill.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.coreskills;

import com.microsoft.semantickernel.builders.SKBuilders;
import com.microsoft.semantickernel.memory.MemoryException;
import com.microsoft.semantickernel.memory.MemoryQueryResult;
import com.microsoft.semantickernel.memory.MemoryRecordMetadata;
import com.microsoft.semantickernel.memory.SemanticTextMemory;
import com.microsoft.semantickernel.orchestration.SKContext;
import com.microsoft.semantickernel.skilldefinition.annotations.DefineSKFunction;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionParameters;
import java.util.List;
import java.util.stream.Collectors;
import javax.annotation.Nonnull;
import reactor.core.publisher.Mono;

/**
 * TextMemorySkill provides a skill to save or recall information from the long or short term.
 * Usage: {@code kernel.ImportSkill(new TextMemorySkill(), "memory");} Example: {@code
 * {{memory.recall "what is the capital of France?" }} => "Paris"}
 */
public class TextMemorySkill {

    /** Name of the context variable used to specify which memory collection to use. */
    public static final String COLLECTION_PARAM = "collection";

    /** Name of the context variable used to specify the information to store. */
    public static final String INFO_PARAM = "info";

    /** Name of the context variable used to specify the input to the skill. */
    public static final String INPUT_PARAM = "input";

    /** Name of the context variable used to specify memory search relevance score. */
    public static final String RELEVANCE_PARAM = "relevance";

    /**
     * Name of the context variable used to specify a unique key associated with stored information.
     */
    public static final String KEY_PARAM = "key";

    /** Name of the context variable used to specify the number of memories to recall */
    public static final String LIMIT_PARAM = "limit";

    /** Default value for the collection name. */
    public static final String DEFAULT_COLLECTION = "generic";

    /** Default value for the relevance score. */
    public static final String DEFAULT_RELEVANCE = "0.75";

    /** Default value for the limit. */
    public static final String DEFAULT_LIMIT = "1";

    /**
     * Key-based lookup for a specific memory.
     *
     * @param collection The collection containing the memory to retrieve
     * @param key The key associated with the memory to retrieve
     * @param context The context containing the semantic memory
     * @return The memory associated with the key
     */
    @DefineSKFunction(description = "Key-based lookup for a specific memory", name = "Retrieve")
    public Mono<String> retrieveAsync(
            @SKFunctionParameters(
                            name = COLLECTION_PARAM,
                            description =
                                    "Memories collection associated with the memory to retrieve",
                            defaultValue = DEFAULT_COLLECTION)
                    String collection,
            @SKFunctionParameters(
                            name = KEY_PARAM,
                            description = "The key associated with the memory to retrieve")
                    String key,
            SKContext context) {

        SemanticTextMemory memory = context.getSemanticMemory();

        if (memory == null) {
            return Mono.error(new RuntimeException("Memory not present"));
        }

        return memory.getAsync(collection, key, false)
                .map(it -> it.getMetadata().getText())
                .defaultIfEmpty("");
    }

    /**
     * Save information to semantic memory.
     *
     * @param info The information to save
     * @param collection Memories collection associated with the information to save
     * @param key The key associated with the information to save
     * @param context The context containing the semantic memory
     * @return The key associated with the saved information
     */
    @DefineSKFunction(description = "Save information to semantic memory", name = "Save")
    public Mono<SKContext> saveAsync(
            @SKFunctionParameters(
                            name = INFO_PARAM,
                            description = "The information to save",
                            defaultValue = "",
                            type = String.class)
                    String info,
            @SKFunctionParameters(
                            name = COLLECTION_PARAM,
                            description =
                                    "Memories collection associated with the information to save",
                            defaultValue = DEFAULT_COLLECTION,
                            type = String.class)
                    String collection,
            @SKFunctionParameters(
                            name = KEY_PARAM,
                            description = "The key associated with the information to save",
                            defaultValue = "",
                            type = String.class)
                    String key,
            SKContext context) {

        SemanticTextMemory memory = context.getSemanticMemory();
        if (memory == null) {
            return Mono.error(
                    new MemoryException(MemoryException.ErrorCodes.UNKNOWN, "Memory not present"));
        }

        return memory.saveInformationAsync(collection, info, key, null, null)
                .map(
                        it -> {
                            context.setVariable(TextMemorySkill.KEY_PARAM, it);
                            return SKBuilders.context()
                                    .with(context.getVariables())
                                    .with(context.getSkills())
                                    .with(context.getSemanticMemory())
                                    .build();
                        });
    }

    /**
     * Semantic search and return up to N memories related to the input text
     *
     * @param input The input text to find related memories for
     * @param collection Memories collection to search
     * @param relevance The relevance score, from 0.0 to 1.0, where 1.0 means perfect match
     * @param limit The maximum number of relevant memories to recall
     * @param context Contains the memory to search
     * @return The list of memories related to the input text
     */
    @DefineSKFunction(
            description = "Semantic search and return up to N memories related to the input text",
            name = "Recall")
    public Mono<List<String>> recallAsync(
            @SKFunctionParameters(
                            name = INPUT_PARAM,
                            description = "The information to recall",
                            defaultValue = "",
                            type = String.class)
                    @Nonnull
                    String input,
            @SKFunctionParameters(
                            name = COLLECTION_PARAM,
                            description =
                                    "Memories collection associated with the information to recall",
                            defaultValue = DEFAULT_COLLECTION,
                            type = String.class)
                    String collection,
            @SKFunctionParameters(
                            name = RELEVANCE_PARAM,
                            description =
                                    "The relevance score, from 0.0 to 1.0, where 1.0 means perfect"
                                            + " match",
                            defaultValue = DEFAULT_RELEVANCE,
                            type = Double.class)
                    double relevance,
            @SKFunctionParameters(
                            name = LIMIT_PARAM,
                            description = "The maximum number of relevant memories to recall",
                            defaultValue = DEFAULT_LIMIT,
                            type = Integer.class)
                    int limit,
            @Nonnull SKContext context) {

        SemanticTextMemory memory = context.getSemanticMemory();
        if (memory == null) {
            return Mono.error(
                    new MemoryException(MemoryException.ErrorCodes.UNKNOWN, "Memory not present"));
        }

        // Validate parameters
        if (collection == null || collection.trim().isEmpty()) {
            collection = context.getVariables().get(TextMemorySkill.COLLECTION_PARAM);
            if (collection == null) collection = DEFAULT_COLLECTION;
        }
        final String _collection = collection;
        final double _relevance = Math.min(1.0, Math.max(0.0, relevance));
        final int _limit = Math.max(1, limit);

        // context.Log.LogTrace("Searching memories in collection '{0}', relevance '{1}'",
        // collection, relevance);

        // Search memory
        return memory.searchAsync(_collection, input, _limit, _relevance, false)
                .flatMap(
                        results -> {
                            List<String> memories =
                                    results.stream()
                                            .map(MemoryQueryResult::getMetadata)
                                            .map(MemoryRecordMetadata::getText)
                                            .collect(Collectors.toList());
                            return Mono.just(memories);
                        });
    }

    /**
     * Remove information from semantic memory
     *
     * @param collection Memories collection associated with the information to remove
     * @param key The key associated with the information to remove
     * @param context Contains the memory to remove
     * @return The reactive stream
     */
    @DefineSKFunction(description = "Remove information from semantic memory", name = "Remove")
    public Mono<Void> removeAsync(
            @SKFunctionParameters(
                            name = COLLECTION_PARAM,
                            description =
                                    "Memories collection associated with the information to remove",
                            defaultValue = DEFAULT_COLLECTION,
                            type = String.class)
                    String collection,
            @SKFunctionParameters(
                            name = KEY_PARAM,
                            description = "The key associated with the information to remove",
                            defaultValue = "",
                            type = String.class)
                    String key,
            SKContext context) {
        SemanticTextMemory memory = context.getSemanticMemory();
        if (memory == null) {
            return Mono.error(
                    new MemoryException(MemoryException.ErrorCodes.UNKNOWN, "Memory not present"));
        }

        return memory.removeAsync(collection, key);
    }
}
